### PR TITLE
fix multi accesslogging not working

### PR DESCRIPTION
**Pull request by zirain**
**Please provide a description of this PR:**

fix: https://github.com/istio/istio/issues/39468

### CODE REVIEW

File: pilot/pkg/model/telemetry.go [hunk 1]
```diff
@@ -461,42 +482,54 @@ func mergeLogs(logs []*tpb.AccessLogging, mesh *meshconfig.MeshConfig, mode tpb.
...
+
 		if len(names) > 0 {
-			providerNames = names
+			providerNames = names.SortedList()
 		}
 
-		if m.Filter != nil {
-			loggingFilter = m.Filter
+		if inScopeKey != m.telemetryKey {
+			inScopeKey = m.telemetryKey
```
[the remark this answers is not in the document]
Comment (zirain) [reply pilot/pkg/model/telemetry.go]: removed, this is unneeded now.

File: pilot/pkg/model/telemetry.go [hunk 2]
```diff
@@ -162,10 +162,15 @@ type tagOverride struct {
 type computedTelemetries struct {
 	telemetryKey
 	Metrics []*tpb.Metrics
-	Logging []*tpb.AccessLogging
+	Logging []*computedAccessLogging
 	Tracing []*tpb.Tracing
 }
 
+type computedAccessLogging struct {
```
[the remark this answers is not in the document]
Comment (zirain) [reply pilot/pkg/model/telemetry.go:172]: update desc

**Review by GregHanson — APPROVED** [review]

File: pilot/pkg/model/telemetry.go [hunk 3]
```diff
@@ -462,41 +485,48 @@ func mergeLogs(logs []*tpb.AccessLogging, mesh *meshconfig.MeshConfig, mode tpb.
...
-		names := getProviderNames(m.Providers)
-		if len(names) > 0 {
-			providerNames = names
+		names := sets.New()
+		for _, p := range m.Logging {
+			subProviders := getProviderNames(p.Providers)
+			names.InsertAll(subProviders...)
+
+			if p.Filter != nil {
+				loggingFilter = p.Filter
```
Comment (howardjohn) [inline pilot/pkg/model/telemetry.go:494]: this seems like we just pick the last one, shouldn't it be per-block?

File: pilot/pkg/model/telemetry.go [hunk 4]
```diff
@@ -462,41 +485,48 @@ func mergeLogs(logs []*tpb.AccessLogging, mesh *meshconfig.MeshConfig, mode tpb.
...
+			}
 		}
 
-		if m.Filter != nil {
-			loggingFilter = m.Filter
+		if len(names) > 0 {
+			providerNames = names.SortedList()
```
Comment (howardjohn) [inline pilot/pkg/model/telemetry.go:499]: nit: if we are just adding this to the inScopeProviders do we even need to sort it? Or convert it out of a set?

File: pilot/pkg/model/telemetry.go (hunk 3 above)
Comment (zirain) [reply pilot/pkg/model/telemetry.go:494]: `Filter` is hard to merged, I have no good idea on it from the begin, so I just keep as before.

Comment (howardjohn) [reply pilot/pkg/model/telemetry.go:494]: Shouldn't it not be merged at all?  Right now its not very safe -- we may try to enabl a provider for only some rare events, but then the filter is ignored and we send all logs?

**Comment by zirain** [comment]
@howardjohn I updated the implement, now `LogFilter` is provider-based, PTAL.

File: releasenotes/notes/39521.yaml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: telemetry
+issue:
+  - https://github.com/istio/istio/issues/39468
+releaseNotes:
+  - |
+    **Fixed** a bug when specifying multi `accessLogging` in Telemetry, only last one is used.
```
Comment (douglas-reid) [inline releasenotes/notes/39521.yaml:8]: this needs to be much clearer, as within a single stanza, you could support multiple access logs and they would all be respected.

File: pilot/pkg/model/telemetry.go [hunk 6]
```diff
@@ -126,10 +126,10 @@ type metricsConfig struct {
...
 type telemetryFilterConfig struct {
 	metricsConfig
-	Provider      *meshconfig.MeshConfig_ExtensionProvider
-	Metrics       bool
-	AccessLogging bool
-	LogsFilter    *tpb.AccessLogging_Filter
+	Provider              *meshconfig.MeshConfig_ExtensionProvider
+	Metrics               bool
+	AccessLogging         bool
+	StackdriverLogsFilter *tpb.AccessLogging_Filter
```
Comment (douglas-reid) [inline pilot/pkg/model/telemetry.go:132]: why is this renamed to `StackdriverLogsFilter`?  The logging filter should be applicable for all logs. Does this PR disable support for basic envoy logs filtering?

File: pilot/pkg/model/telemetry.go [hunk 7]
```diff
@@ -424,12 +456,19 @@ func (t *Telemetries) telemetryFilters(proxy *Proxy, class networking.ListenerCl
 		}
 		_, logging := tml[k]
 		_, metrics := tmm[k]
+
+		var logsFilter *tpb.AccessLogging_Filter
+		switch p.Provider.(type) {
+		case *meshconfig.MeshConfig_ExtensionProvider_Stackdriver:
```
Comment (douglas-reid) [inline pilot/pkg/model/telemetry.go:462]: this is not right. the filter should not be only taken for Stackdriver.  why is this change being made?

File: pilot/pkg/model/telemetry_test.go [hunk 8]
```diff
@@ -276,6 +277,96 @@ func TestAccessLogging(t *testing.T) {
 			},
 		},
 	}
+	multiAccessLogging := &tpb.Telemetry{
+		AccessLogging: []*tpb.AccessLogging{
```
Comment (douglas-reid) [inline pilot/pkg/model/telemetry_test.go:281]: we need tests cases for a single providers stanza with multiple provider names.

File: pilot/pkg/model/telemetry_test.go [hunk 9]
```diff
@@ -380,6 +471,22 @@ func TestAccessLogging(t *testing.T) {
 			nil,
 			[]string{},
 		},
+		{
+			"server - mutli filters",
```
Comment (douglas-reid) [inline pilot/pkg/model/telemetry_test.go:475]:
```suggestion
			"server - multi filters",
```

**Review by douglas-reid — CHANGES_REQUESTED** [review]

File: pilot/pkg/model/telemetry.go (hunk 7 above)
Comment (zirain) [reply pilot/pkg/model/telemetry.go:462]: in function `buildHTTPTelemetryFilter`, there's a comment `Only prometheus and SD supported currently`

File: pilot/pkg/model/telemetry.go (hunk 6 above)
Comment (zirain) [reply pilot/pkg/model/telemetry.go:132]: no, basic envoy logs are using `AccessLogging`, right now only SD need `telemetryFilters`.

File: pilot/pkg/model/telemetry.go (hunk 7 above)
Comment (zirain) [reply pilot/pkg/model/telemetry.go:462]: you're right, updated the implements.

File: pilot/pkg/model/telemetry.go [hunk 10]
```diff
@@ -449,58 +482,66 @@ func (t *Telemetries) telemetryFilters(proxy *Proxy, class networking.ListenerCl
...
 	providerNames := mesh.GetDefaultProviders().GetAccessLogging()
+	filters := map[string]*tpb.AccessLogging_Filter{}
 	for _, m := range logs {
-		names := getProviderNames(m.Providers)
-		if len(names) > 0 {
-			providerNames = names
+		names := sets.New()
+		for _, p := range m.Logging {
+			subProviders := getProviderNames(p.Providers)
+			names.InsertAll(subProviders...)
```
Comment (j2gg0s) [inline pilot/pkg/model/telemetry.go:501]: If we insert any AccessLogging's providers to `inScopeProviders`, do we still need check `isScopeProviders.Containers` in L524?

File: releasenotes/notes/39521.yaml [hunk 11]
```diff
@@ -0,0 +1,8 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: telemetry
+issue:
+  - https://github.com/istio/istio/issues/39468
+releaseNotes:
+  - |
+    **Fixed** a bug when specifying multi `accessLogging` in Telemetry within a single stanza, now they would all be respected.
```
Comment (douglas-reid) [inline releasenotes/notes/39521.yaml:8]:
```suggestion
    **Fixed** a bug that would ignore some configuration when specifying multiple `accessLogging` in Telemetry resources within a single stanza. With this fix, all provided access logging configuration within a single stanza of Telemetry resource are respected.
```

File: pilot/pkg/networking/core/v1alpha3/accesslog.go [hunk 12]
```diff
@@ -167,28 +167,33 @@ func (b *AccessLogBuilder) setTCPAccessLog(push *model.PushContext, proxy *model
...
+			continue
+		}
 
-	for _, p := range spec.Providers {
 		var al *accesslog.AccessLog
-		switch prov := p.Provider.(type) {
+		switch prov := spec.Provider.Provider.(type) {
 		case *meshconfig.MeshConfig_ExtensionProvider_EnvoyFileAccessLog:
 			// For built-in provider, fallback to Mesh Config for formatting options.
-			if p.Name == defaultEnvoyAccessLogProvider {
```
[the remark this answers is not in the document]
Comment (zirain) [reply pilot/pkg/networking/core/v1alpha3/accesslog.go:191]: we have discuss this in https://github.com/istio/istio/pull/39606

File: pilot/pkg/model/telemetry.go [hunk 13]
```diff
@@ -449,58 +482,66 @@ func (t *Telemetries) telemetryFilters(proxy *Proxy, class networking.ListenerCl
...
+
+			for _, prov := range subProviders {
+				// TODO: consider pick the non-nil?
```
[the remark this answers is not in the document]
Comment (zirain) [reply pilot/pkg/model/telemetry.go:504]: thinking about the following configuration:
```yaml
apiVersion: telemetry.istio.io/v1alpha1
kind: Telemetry
metadata:
  name: mutli-accessLogging
  namespace: istio-system
spec:
  accessLogging:
  - providers:
    - name: envoy
    - name: envoy-json
    filter: 
      expression: response.code >= 400
  - providers:
    - name: envoy-json
```

what's  `envoy-json`'s filter should be? ` response.code >= 400` or `nil`?

[the remark this answers is not in the document]
Comment (zirain) [reply pilot/pkg/model/telemetry.go:505]: the root cause is how to merge `filter`.

File: pilot/pkg/model/telemetry.go [hunk 14]
```diff
@@ -449,58 +482,66 @@ func (t *Telemetries) telemetryFilters(proxy *Proxy, class networking.ListenerCl
 }
 
 // mergeLogs returns the set of providers for the given logging configuration.
```
[the remark this answers is not in the document]
Comment (zirain) [reply pilot/pkg/model/telemetry.go]: sounds good

**Comment by zirain** [comment]
@douglas-reid @howardjohn refactor base on https://github.com/istio/istio/pull/39751, please review again.

do we need backport this to v1.13/v.14, cause there's a warn validating.
https://github.com/istio/istio/blob/d9762744d63553d6929a2b548ae8300cd3d78eb4/pkg/config/validation/validation.go#L3665

**Review by GregHanson — APPROVED** [review]

File: pilot/pkg/networking/core/v1alpha3/accesslog.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (douglas-reid) [reply pilot/pkg/networking/core/v1alpha3/accesslog.go]: I don't understand.  Can you please clarify?  That PR was about falling back to MeshConfig. This logic is replacing the provider completely. That will break user expectations completely.

[the remark this answers is not in the document]
Comment (zirain) [reply pilot/pkg/networking/core/v1alpha3/accesslog.go]: I got what your said, do you means, we should only fallback to MeshConfig when `p.Name == "envoy"` and `prov.EnvoyFileAccessLog.Path==""`?

**Comment by howardjohn** [comment]
> @douglas-reid @howardjohn refactor base on #39751, please review again.
> 
> do we need backport this to v1.13/v.14, cause there's a warn validating.
> 
> https://github.com/istio/istio/blob/d9762744d63553d6929a2b548ae8300cd3d78eb4/pkg/config/validation/validation.go#L3665

I don't think we need to, fine with just on master.

File: pilot/pkg/networking/core/v1alpha3/accesslog.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (douglas-reid) [reply pilot/pkg/networking/core/v1alpha3/accesslog.go]: let's leave this for now. i'm worried that people will create their own provider named `envoy` and this will be a mess, but we can address that later.

File: pilot/pkg/networking/core/v1alpha3/accesslog.go [hunk 15]
```diff
@@ -127,9 +127,9 @@ func newAccessLogBuilder() *AccessLogBuilder {
 
 func (b *AccessLogBuilder) setTCPAccessLog(push *model.PushContext, proxy *model.Proxy, tcp *tcp.TcpProxy, class networking.ListenerClass) {
 	mesh := push.Mesh
-	cfg := push.Telemetry.AccessLogging(push, proxy, class)
+	cfgs := push.Telemetry.AccessLogging(push, proxy, class)
 
-	if cfg == nil {
+	if cfgs == nil {
```
Comment (douglas-reid) [inline pilot/pkg/networking/core/v1alpha3/accesslog.go:132]: curious: should we also check `len(cfgs)`? naively, it seems like `len(cfgs) == 0` might be appropriate (here and below) and cover the `nil` case too.

**Review by douglas-reid — APPROVED** [review]
one question remaining. otherwise looks reasonable. thanks for tracking this down and your patience in reviews.

File: pilot/pkg/model/telemetry.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (douglas-reid) [reply pilot/pkg/model/telemetry.go]: Do we want to document how such a config might work in the API docs?  That would be helpful.

File: pilot/pkg/networking/core/v1alpha3/accesslog.go (hunk 15 above)
Comment (zirain) [reply pilot/pkg/networking/core/v1alpha3/accesslog.go:132]: we can improve this later, but for now keep the behavior as before, more details can be found in following test case:
```golang
                {
			"empty",
			nil,
			networking.ListenerClassSidecarOutbound,
			sidecar,
			nil,
			nil, // No Telemetry API configured, fall back to legacy mesh config setting
		},
```

File: pilot/pkg/model/telemetry.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (zirain) [reply pilot/pkg/model/telemetry.go]: plan to update all the new things(`mode`, `filter` etc)

**Comment by zirain** [comment]
kindly ping @istio/wg-networking-maintainers-pilot

**Review by howardjohn — APPROVED** [review]
